### PR TITLE
Add experience rewards

### PR DESCRIPTION
**Pull request by francisco-leal**
Notion task: https://www.notion.so/talentprotocol/Bounty-Rewards-Tab-288a5cca3abf44469c97ead6f23bafb9?d=0808fe22e84541ef8f54abe012688e94

- Adds several new database structures:
  - ExperienceRewards: the different type of rewards that we have for ExperiencePoints
  - ExperienceRewardClaims: Keep track of which rewards have been claimed and by who
  - MerchCode: to hold all different merch codes related to rewards

### CODE REVIEW

File: config/routes/public_api_routes.rb [hunk 1]
```diff
@@ -18,6 +18,9 @@ def self.extended(router)
               get :of_user, on: :collection
             end
             resources :connections, only: [:index]
+            resources :experience_rewards, only: [:index] do
```
Comment (francisco-leal) [inline config/routes/public_api_routes.rb:21]: this creates get `api/v1/experience_rewards` and POST/PUT `api/v1/experience_rewards/:uuid/claim`

File: app/controllers/api/v1/public_api/experience_rewards_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+class API::V1::PublicAPI::ExperienceRewardsController < API::V1::PublicAPI::APIController
+  def index
+    rewards = ExperienceReward.active.order(:cost)
+
+    render json: {rewards: API::ExperienceRewardBlueprint.render_as_json(rewards, view: :normal, current_user: current_user)}, status: :ok
+  end
+
+  def claim
+    reward = ExperienceReward.find_by(uuid: params[:experience_reward_id])
```
Comment (RubenSousaDinis) [inline app/controllers/api/v1/public_api/experience_rewards_controller.rb:9]: We already have a catch here: https://github.com/talentprotocol/web-dapp/blob/b34915e15ed8b354f49b3c2da636d38d1889ca48/app/controllers/api/v1/public_api/api_controller.rb#L60

You can remove the if/else.

```suggestion
    reward = ExperienceReward.find_by!(uuid: params[:experience_reward_id])
```

File: app/controllers/api/v1/public_api/experience_rewards_controller.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    if reward.nil?
+      render json: {error: "Reward not found"}, status: :not_found
+    else
+      service = Rewards::Claim.new(user: current_user, reward: reward)
+      result = service.call
+
+      if result
+        render json: {success: result}, status: :ok
```
Comment (RubenSousaDinis) [inline app/controllers/api/v1/public_api/experience_rewards_controller.rb:18]: This approach is not following our Public API standard. We should update it to be consistent. Creating the spec/doc will help to follow the standard

Comment (RubenSousaDinis) [reply app/controllers/api/v1/public_api/experience_rewards_controller.rb:18]: The index endpoint is following the standard.

File: app/controllers/api/v1/public_api/experience_rewards_controller.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+class API::V1::PublicAPI::ExperienceRewardsController < API::V1::PublicAPI::APIController
+  def index
+    rewards = ExperienceReward.active.order(:cost)
+
+    render json: {rewards: API::ExperienceRewardBlueprint.render_as_json(rewards, view: :normal, current_user: current_user)}, status: :ok
+  end
+
+  def claim
```
Comment (RubenSousaDinis) [inline app/controllers/api/v1/public_api/experience_rewards_controller.rb:8]: Missing authentication.  You can add the following before actions only for the claim endpoint

```
before_action :internal_only
before_action :authenticated_only
```

**Review by RubenSousaDinis — CHANGES_REQUESTED** [review]
Some comments to discuss

File: app/controllers/api/v1/public_api/experience_rewards_controller.rb (hunk 3 above)
Comment (francisco-leal) [reply app/controllers/api/v1/public_api/experience_rewards_controller.rb:18]: This endpoint already has specs 😄 where can I read the standard?

File: app/controllers/api/v1/public_api/experience_rewards_controller.rb (hunk 2 above)
Comment (francisco-leal) [reply app/controllers/api/v1/public_api/experience_rewards_controller.rb:9]: This allows me to send a specific message on this endpoint to help the frontend show a more meaningful message. (the message itself will still be improved)

File: app/services/rewards/claim.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+      @reward = reward.becomes(reward.type.constantize)
+    end
+
+    def call
+      # reward out of stock
+      return if @reward.stock <= 0 || @reward.claimed?(@user)
+      # user does not have enough experience
+      return if @user.experience_points.sum(:amount) < @reward.cost
+
+      ActiveRecord::Base.transaction do
```
Comment (RubenSousaDinis) [inline app/services/rewards/claim.rb:16]: Missing lock with stock check

**Review by RubenSousaDinis — COMMENTED** [review]
Some comments to discuss

File: app/services/rewards/claim.rb (hunk 5 above)
Comment (RubenSousaDinis) [reply app/services/rewards/claim.rb:11]: https://api.rubyonrails.org/classes/ActiveRecord/Locking/Pessimistic.html. You can check some ways here to simplify the code.

```suggestion
 account.with_lock do
    # This block is called within a transaction,
    # account is already locked.
    account.balance -= 100
    account.save!
end

```

File: app/services/rewards/claim.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    def initialize(user:, reward:)
+      @user = user
+      # ensure we are using the subclass
+      @reward = reward
+    end
+
+    def call
+      ActiveRecord::Base.transaction do
+        # reward out of stock
+        reward = @reward.type.constantize.lock("FOR UPDATE NOWAIT").find(@reward.id)
```
Comment (RubenSousaDinis) [inline app/services/rewards/claim.rb:13]: Let's try to be consistent and use private attr_readers. We're using this approach in most of the services

File: db/migrate/20230905111753_create_merch_codes.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class CreateMerchCodes < ActiveRecord::Migration[7.0]
+  def change
+    create_table :merch_codes do |t|
+      t.references :experience_reward, null: false, foreign_key: true
+      t.string :code
```
[the remark this answers is not in the document]
Comment (francisco-leal) [reply db/migrate/20230905111753_create_merch_codes.rb:5]: No need, not that relevant

**Review by RubenSousaDinis — APPROVED** [review]

**Pull request merged**